builtins: add ulid functionality

lopezator opened the pull request:
This patch implements the builtin `gen_ulid`.

Release justification: low-risk update to new functionality
Release note (sql change): The `gen_ulid` builtin is now
available for use.

Closes #24733

lopezator committed:
builtins: add ulid functionality

This patch implements `gen_ulid` and
`uuid_ulid_to_string` builtins.

Release justification: low-risk update to new functionality
Release note (sql change): The `gen_ulid` and
`uuid_ulid_to_string` builtins are now available for use.

Closes #24733

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62440)
<!-- Reviewable:end -->

lopezator commented:
ulid spec and recommended tools here: https://github.com/ulid/spec

I have some questions about the PR:

1. Do we need a new type (like UUID) or can we use just a string?
2. I'm not sure how to test this functionality. I did  a manual test as follow to verify it's working:

```
demo@127.0.0.1:26257/movr> CREATE TABLE foo (id UUID NOT NULL DEFAULT gen_random_uuid(), ulid STRING NOT NULL DEFAULT gen_ulid(), text TEXT);
CREATE TABLE

Time: 7ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/movr> INSERT INTO foo (text) VALUES ('foo');
INSERT 1

Time: 7ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/movr> SELECT * from foo;
                   id                  |            ulid            | text
---------------------------------------+----------------------------+-------
  2c0f8e86-48ac-4c7f-b0ed-c4de21c67a81 | 01F1EJJQ9KKRYQK4TSXG69NMZ7 | foo
(1 row)

Time: 1ms total (execution 1ms / network 0ms)
```

lopezator commented:
@jordanlewis tutorial https://www.youtube.com/watch?v=tPLI6plgbfA here helped getting this started, thank you!

lopezator commented:
At this point to be able to follow with this I have a question:

I suspect that we need a new `oidext` for ULIDs:

```
T_ulid:       "ULID",
T__ulid:      "_ULID",
``` 

But do we need a new family too? Or would you use `types.UUIDFamily` for `T_ulid` also?

CC/ @jordanlewis @dt @bdarnell

jordanlewis commented:
Hi @lopezator, sorry for the delay here! Thanks for the hard work and we'll review this soon.

lopezator commented:
> Hi @lopezator, sorry for the delay here! Thanks for the hard work and we'll review this soon.

You are welcome! 

I have an alternative implementation on other branch adding a new ULIDFamily and a new oidext identificator for ulid. I have an err on that implemention with DatumVec conv and generated code though.

I was just waiting for some feedback/guidance before proceeding.

Best,

David.

RichardJCai commented:
> * Can we just use a string like in this initial proposal? (I suppose we can't)
> *

I'm somewhat hesitant of adding a new ULID data type since we generally don't add new types.

I'm not familiar with ULIDs but is it possible to generate ULIDs in the format of a UUID? 

https://github.com/valohai/ulid2 this lib supports a generate_ulid_as_uuid function, maybe we can do something similar?

lopezator commented:
> I'm not familiar with ULIDs but is it possible to generate ULIDs in the format of a UUID?

I suppose we can, but how would you print that ULID as UUID (UUID type) back as an ULID string? Adding a new variant or something like that to [uuid package](https://github.com/cockroachdb/cockroach/blob/f80ec9c292d43676d739fffabc847fb8b3195396/pkg/util/uuid/uuid.go#L136)? Sounds hacky.

I've found this related packages also:

https://github.com/ahawker/ulid  
https://github.com/yaslab/ULID.swift

RichardJCai commented:
> > I'm not familiar with ULIDs but is it possible to generate ULIDs in the format of a UUID?
> 
> I suppose we can, but how would you print that ULID as UUID (UUID type) back as an ULID string? Adding a new variant or something like that to [uuid package](https://github.com/cockroachdb/cockroach/blob/f80ec9c292d43676d739fffabc847fb8b3195396/pkg/util/uuid/uuid.go#L136)? Sounds hacky.
> 
> I've found this related packages also:
> 
> https://github.com/ahawker/ulid
> https://github.com/yaslab/ULID.swift

One option is we can provide a builtin to convert the UUID to a ULID string.

lopezator commented:
> One option is we can provide a builtin to convert the UUID to a ULID string.

That's a great idea! Thank you for the hint.

I suppose then, that something like this could work: 

https://play.golang.org/p/UQ0sXUPLz9s

Will update the proposal based on that and see if it works.

rafiss reviewed (changes requested):
this is looking good!

can you add some tests to `pkg/sql/logictest/testdata/logic_test/builtin_function`

obviously, the tests won't be super deep since it's a random function, but look at how `gen_random_uuid` is tested there


also, i wonder if we should just copy the ulid.go and ulid_test.go from https://github.com/oklog/ulid and keep it in this repo, for ease of modification. it's under the apache license, so we'd have to preserve the headers and reference the license

see pkg/util/uuid/codec.go for a case where we did that for an MIT-licensed project

rafiss commented:
Oops i left these comments before seeing that @RichardJCai reviewed this too!

i agree with that point -- it's better if we can avoid introducing a new type.

lopezator commented:
I updated the code addressing your comments.

1. Added two builtin functions to avoid creating a type as @RichardJCai suggested:

  - gen_random_ulid
  - uuid_ulid_to_string

2. Copied the `ulid.go` and `ulid_test.go` files to ulid package as @rafiss suggested.
3. Added tests for both functions

Manual testing worked fine:

```
demo@127.0.0.1:26257/movr> CREATE TABLE foo (ulid UUID NOT NULL DEFAULT gen_random_ulid(), text TEXT);
CREATE TABLE

Time: 6ms total (execution 6ms / network 0ms)

demo@127.0.0.1:26257/movr> INSERT INTO foo (text) VALUES ('foo');
INSERT 1

Time: 7ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/movr> select * from foo;
                   ulid                 | text
--------------------------------------+-------
  0178951c-b665-30a7-19a2-a18999834858 | foo
(1 row)

Time: 2ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/movr> select ulid from foo;
                  ulid
----------------------------------------
  0178951c-b665-30a7-19a2-a18999834858
(1 row)

Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/movr> select uuid_ulid_to_string(ulid) from foo;
     uuid_ulid_to_string
------------------------------
  01F2AHSDK562KHK8N1H6CR6J2R
(1 row)

Time: 1ms total (execution 1ms / network 0ms)
```
Thank you for the help!

lopezator commented:
Open for suggestions regarding to the naming of the builtin functions.

lopezator commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -609,6 +610,41 @@ var builtins = map[string]builtinDefinition{
+			Types:      tree.ArgTypes{},
+			ReturnType: tree.FixedReturnType(types.Uuid),
+			Fn: func(_ *tree.EvalContext, _ tree.Datums) (tree.Datum, error) {
+				t := time.Now().UnixNano()
```
Should we use here `time.Now().UTC().UnixNano()` instead?

RichardJCai replied:
Don't believe that's necessary.

RichardJCai commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -609,6 +610,41 @@ var builtins = map[string]builtinDefinition{
+		},
+	),
+
+	"uuid_ulid_to_string": makeBuiltin(defProps(),
```
I was thinking something like `uuid_to_ulid` or `uid_to_ulid_string` makes more sense to me.

lopezator replied:
The intention was to make clear that you were casting an "uuid-formmated" ulid, but I get the point. If you feel that we won't need this name in the future I would go for the short form `uuid_to_ulid`.

RichardJCai replied:
Yeah I understand what you were going for, I think as long as were clear in the description of the function, it's fine. Technically any UUID can be turned into a ULID but users should know what the function is used for by the description.

I think `uuid_to_ulid` is fine.

lopezator replied:
Cool! I'll got for it then.

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -1928,6 +1928,16 @@ SELECT length(gen_random_uuid()::BYTES), gen_random_uuid() = gen_random_uuid()
+query TTTTTT
+SELECT uuid_ulid_to_string('0178951c-b665-30a7-19a2-a18999834858')
+----
+01F2AHSDK562KHK8N1H6CR6J2R
```
I think we should test roundtripping this back to a UUID, we can add another builtin so we can go both ways for the test.

lopezator replied:
Makes sense, I'll add `ulid_to_uuid` here based on the previous comment.

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -1928,6 +1928,16 @@ SELECT length(gen_random_uuid()::BYTES), gen_random_uuid() = gen_random_uuid()
+SELECT uuid_ulid_to_string('0178951c-b665-30a7-19a2-a18999834858')
+----
+01F2AHSDK562KHK8N1H6CR6J2R
+
```
Also can we add a test for monotonicity.

lopezator replied:
Not sure how to do this in a clean way only using this input/output text file, without writing any specific test (maybe I should do this)

I've been looking here:

https://github.com/oklog/ulid/blob/0d4fda9d6345755e157a256fd33d48556c5f4a7a/ulid_test.go#L520-L560

Any ideas?

RichardJCai replied:
There's this neat `let` feature that lets you store the output for a command.
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/sequences_regclass#L100

We can create two ULIDs and use a string comparison to ensure the ordering is correct.

lopezator replied:
This is great info, I updated the code making use of let :smile:

RichardJCai reviewed (changes requested):
This is awesome! Thanks!

Just a couple of comments but overall this looks really good.

lopezator commented:
> Don't believe that's necessary.

Not sure want you meant here, could you expand this please? Were you talking about the `t` variable? I am using it below. 

I don't now why, but GitHub doesn't allow me to answer to that specific comment inline :shrug:

lopezator commented:
> > Don't believe that's necessary.
> 
> Not sure want you meant here, could you expand this please? Were you talking about the `t` variable? I am using it below.
> 
> I don't now why, but GitHub doesn't allow me to answer to that specific comment inline shrug

Nevermind I suppose that the comment was replicated in two places, I just got it lol:)

lopezator commented:
Thanks a lot for the feedback! I'll get hands to work right now :).

lopezator commented:
Code updated, PTAL @RichardJCai

RichardJCai commented:
> Code updated, PTAL @RichardJCai

Nice, looks like we're almost done here.
By the way, are you able to do a `make bazel-generate`
Bazel CI is unhappy here since it's missing the new files

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62440)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @jordanlewis, @lopezator, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1945 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXts149DlVyuIKBEYmW:-MXts14A9it4ZKWrQ3Bu:b-wowl4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1945)):*
> ```
> 0178951c-b665-30a7-19a2-a18999834858
> 0178951c-b665-30a7-19a2-a18999834858
> 01F2AHSDK562KHK8N1H6CR6J2R
> ```

these should all be on the same line, each separated by 2 spaces

```
01F2AHSDK562KHK8N1H6CR6J2R  0178951c-b665-30a7-19a2-a18999834858  0178951c-b665-30a7-19a2-a18999834858  01F2AHSDK562KHK8N1H6CR6J2R
```
___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1954 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrtgi5McNVWqZ969W:-MXtrtgi5McNVWqZ969X:b-6x8dt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1954)):*
> ```
> 
> query TTTT
> select '$ulid1' < '$ulid2',
> ```

nitpick: capitalize SELECT
___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1957 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrrGG3ZIofY3RlKxZ:-MXtrrGG3ZIofY3RlKx_:bbxwif9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1957)):*
> ```
>        '$ulid1' > '$ulid2',
>        uuid_to_ulid('$ulid1') < uuid_to_ulid('$ulid2'),
>        uuid_to_ulid('$ulid1') > uuid_to_ulid('$ulid2'),
> ```

remove the trailing comma to fix this
___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1958 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrw-7E_b3H6XecL2w:-MXtrw-7E_b3H6XecL2x:b-sh6jux) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1958)):*
> ```
>        uuid_to_ulid('$ulid1') < uuid_to_ulid('$ulid2'),
>        uuid_to_ulid('$ulid1') > uuid_to_ulid('$ulid2'),
> ```

nitpick: remove the empty line
___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtsCOw47yJBbrnMbbf:-MXtsCOw47yJBbrnMbbg:bs4mayv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1963)):*
> ```
> false
> true
> false
> ```

these should be on the same line, separated by two spaces

```
true  false  true  false
```


<!-- Sent from Reviewable.io -->

lopezator commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62440)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @jordanlewis, @lopezator, and @RichardJCai)
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1945 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXts149DlVyuIKBEYmW:-MXts14A9it4ZKWrQ3Bu:b-wowl4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1945)):_
> 
> > ```
> > 0178951c-b665-30a7-19a2-a18999834858
> > 0178951c-b665-30a7-19a2-a18999834858
> > 01F2AHSDK562KHK8N1H6CR6J2R
> > ```
> 
> these should all be on the same line, each separated by 2 spaces
> 
> ```
> 01F2AHSDK562KHK8N1H6CR6J2R  0178951c-b665-30a7-19a2-a18999834858  0178951c-b665-30a7-19a2-a18999834858  01F2AHSDK562KHK8N1H6CR6J2R
> ```
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1954 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrtgi5McNVWqZ969W:-MXtrtgi5McNVWqZ969X:b-6x8dt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1954)):_
> 
> > ```
> > 
> > query TTTT
> > select '$ulid1' < '$ulid2',
> > ```
> 
> nitpick: capitalize SELECT
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1957 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrrGG3ZIofY3RlKxZ:-MXtrrGG3ZIofY3RlKx_:bbxwif9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1957)):_
> 
> > ```
> >        '$ulid1' > '$ulid2',
> >        uuid_to_ulid('$ulid1') < uuid_to_ulid('$ulid2'),
> >        uuid_to_ulid('$ulid1') > uuid_to_ulid('$ulid2'),
> > ```
> 
> remove the trailing comma to fix this
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1958 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtrw-7E_b3H6XecL2w:-MXtrw-7E_b3H6XecL2x:b-sh6jux) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1958)):_
> 
> > ```
> >        uuid_to_ulid('$ulid1') < uuid_to_ulid('$ulid2'),
> >        uuid_to_ulid('$ulid1') > uuid_to_ulid('$ulid2'),
> > ```
> 
> nitpick: remove the empty line
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MXtsCOw47yJBbrnMbbf:-MXtsCOw47yJBbrnMbbg:bs4mayv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ff8a8bf3dc31da9b94e9e2f38936adfc249cfa8/pkg/sql/logictest/testdata/logic_test/builtin_function#L1963)):_
> 
> > ```
> > false
> > true
> > false
> > ```
> 
> these should be on the same line, separated by two spaces
> 
> ```
> true  false  true  false
> ```

Thank you for the comments, changes applied.

By the way, is there any documentation somewhere for the testdata files syntax?

lopezator commented on pkg/util/ulid/ulid_test.go:
```diff
@@ -0,0 +1,868 @@
+}
+
+func TestNew(t *testing.T) {
+	t.Parallel() // SAFE FOR TESTING (this comment is for the linter)
```
I had to adapt both `ulid.go` and `ulid_test.go` to met linter rules, but here I cheated.

As `t.Parallel` is forbidden by the linter, I added  exceptions for them.

- Are you ok with this exceptions here? 
- Is best to just remove every `t.Parallel`? 
- Do I have to adapt the code to use `sync.WorkGroups` and rewrite all the tests?

RichardJCai replied:
Personally I'm fine with this. Doing the least amount of work here makes sense, probably worse off to change the tests more.

lopezator replied:
👌

lopezator commented:
I suspect that monotonicity isn't working properly at the millisecond level for the ulids stored in the `let` vars `ulid1` and `ulid2`.

But trying the same code in a separate project (two ulids inside the same millisecond) and then sorting the ulids inside cockroachdb worked fine.

Something related to the execution order? Race condition? 

I have no idea, I will need to investigate further 🤷‍♂️.

RichardJCai commented:
> I suspect that monotonicity isn't working properly at the millisecond level for the ulids stored in the `let` vars `ulid1` and `ulid2`.
> 
> But trying the same code in a separate project (two ulids inside the same millisecond) and then sorting the ulids inside cockroachdb worked fine.
> 
> Something related to the execution order? Race condition?
> 
> I have no idea, I will need to investigate further 🤷‍♂️.

Hmm that's interesting, I'm not sure why execution order would be an issue here.
I think the issue is most likely that `Within the same millisecond, sort order is not guaranteed`  

Try putting a `crdb_internal.pg_sleep(0.001)` between, I think that will likely solve it.

lopezator commented:
> > I suspect that monotonicity isn't working properly at the millisecond level for the ulids stored in the `let` vars `ulid1` and `ulid2`.
> > But trying the same code in a separate project (two ulids inside the same millisecond) and then sorting the ulids inside cockroachdb worked fine.
> > Something related to the execution order? Race condition?
> > I have no idea, I will need to investigate further man_shrugging.
> 
> Hmm that's interesting, I'm not sure why execution order would be an issue here.
> I think the issue is most likely that `Within the same millisecond, sort order is not guaranteed`
> 
> Try putting a `crdb_internal.pg_sleep(0.001)` between, I think that will likely solve it.

What confuses me is this sentence in the [README](https://github.com/oklog/ulid) of `oklog/ulid`:

> Monotonic sort order (correctly detects and handles the same millisecond)

I suppose that I am missing something, but doesn't this mean that this should work without needing to sleep?

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -1928,6 +1928,36 @@ SELECT length(gen_random_uuid()::BYTES), gen_random_uuid() = gen_random_uuid()
+SELECT gen_random_ulid()
+
+# We need this to ensure that the ULID sorting within the same millisecond works properly
+crdb_internal.pg_sleep(0.001)
```
Need to have a `statement ok` above this so the logic test knows to execute the following statement.

By the way are you able to run these logictests locally? To run this one specifically you can do `make testlogic FILES=builtin_function` from the root of the repo.

lopezator replied:
I had no idea TBH. This helped a lot, thanks!

lopezator commented:
It's green finally, thank @RichardJCai

RichardJCai reviewed (approved):
Thanks for your hard work @lopezator! Looks great just minor nits on the logic test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-:-MY6Is2b6MntuWQpwFJ8:bn1po9x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @jordanlewis, @lopezator, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1947 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MY6INYWDUNvyIAfgDhv:-MY6INYWDUNvyIAfgDhw:bxpntsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e3556c783483bdf2cb9e750c5fe31ed6b32564/pkg/sql/logictest/testdata/logic_test/builtin_function#L1947)):*
> ```
> SELECT gen_random_ulid()
> 
> # We need this to ensure that the ULID sorting within the same millisecond works properly
> ```

nit: We're not actually sorting in the same millisecond, this is to ensure at least a millisecond passes between gen_ulid. Also please end comment in period.
___
*[pkg/sql/logictest/testdata/logic_test/builtin_function, line 2495 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MY6I_lr7u33wqY7vYYQ:-MY6I_lr7u33wqY7vYYR:bsaij6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e3556c783483bdf2cb9e750c5fe31ed6b32564/pkg/sql/logictest/testdata/logic_test/builtin_function#L2495)):*
> ```
> SELECT crdb_internal.check_consistency(true, '\x03', '\x02')
> 
> query ITT
> ```

Was this test for debugging? If so please remove


<!-- Sent from Reviewable.io -->

rafiss commented:
also, to answer your earlier question:

> By the way, is there any documentation somewhere for the testdata files syntax?

yes, this is documented in the main comment of pkg/sql/logictest/logic.go

lopezator commented:
> Thanks for your hard work @lopezator! Looks great just minor nits on the logic test.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-:-MY6Is2b6MntuWQpwFJ8:bn1po9x)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @dt, @jordanlewis, @lopezator, @rafiss, and @RichardJCai)
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 1947 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MY6INYWDUNvyIAfgDhv:-MY6INYWDUNvyIAfgDhw:bxpntsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e3556c783483bdf2cb9e750c5fe31ed6b32564/pkg/sql/logictest/testdata/logic_test/builtin_function#L1947)):_
> 
> > ```
> > SELECT gen_random_ulid()
> > 
> > # We need this to ensure that the ULID sorting within the same millisecond works properly
> > ```
> 
> nit: We're not actually sorting in the same millisecond, this is to ensure at least a millisecond passes between gen_ulid. Also please end comment in period.
> 
> _[pkg/sql/logictest/testdata/logic_test/builtin_function, line 2495 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/62440#-MY6I_lr7u33wqY7vYYQ:-MY6I_lr7u33wqY7vYYR:bsaij6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e3556c783483bdf2cb9e750c5fe31ed6b32564/pkg/sql/logictest/testdata/logic_test/builtin_function#L2495)):_
> 
> > ```
> > SELECT crdb_internal.check_consistency(true, '\x03', '\x02')
> > 
> > query ITT
> > ```
> 
> Was this test for debugging? If so please remove

Fixed. Thanks to you for the fast feedback and the help provided!

RichardJCai commented:
Once again thanks for your hard work @lopezator !

bors r+

lopezator commented:
> Once again thanks for your hard work @lopezator !
> 
> bors r+

You are welcome, happy to help.

bdarnell commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -609,6 +610,65 @@ var builtins = map[string]builtinDefinition{
+			Types:      tree.ArgTypes{},
+			ReturnType: tree.FixedReturnType(types.Uuid),
+			Fn: func(_ *tree.EvalContext, _ tree.Datums) (tree.Datum, error) {
+				entropy := ulid.Monotonic(rand.New(rand.NewSource(timeutil.Now().UnixNano())), 0)
```
This creates a new random source from the current time every call to `gen_random_ulid`. This is not an effective source of entropy. It should either use a cryptographic source of randomness (as we do for UUIDs) or persist a PRNG somewhere for repeated use. (We should strongly favor the use of cryptographic randomness).

RichardJCai replied:
Gotcha, I'm not too familiar with RNG but in this case, instead of using the current time, would using `crypto/rand` to also generate the seed make this acceptable?

https://golang.org/pkg/crypto/rand/ `Package rand implements a cryptographically secure random number generator.`

Specifically using https://golang.org/pkg/crypto/rand/#Int to generate the seed for the source?

bdarnell replied:
Using a cryptographic RNG to seed a PRNG is better than just seeding a PRNG with the current time, but in general it's better to just use the cryptographic RNG directly (this is what UUIDs do). Can't you just pass `crypto/rand.Reader` to `ulid.Monotonic`?

lopezator replied:
@bdarnell Let's see if I understood you correctly.

https://github.com/cockroachdb/cockroach/pull/65768

Thank you for the review.

lopezator replied:
Also thank you @RichardJCai for all the help!

alvitawa commented:
Would it be possible to add an optional timestamp argument? This would to write migrations to ULID based on existing date column.
